### DISCUSSION TITLE
useWebNotification not working on chrome mobile

### DISCUSSION THREAD

**Discussion by ibnusyawall** [vueuse/vueuse#3978] [Q&A]
I have grant permission of the notification request but notification not showing on mobile, works on dekstop only

**Comment by ilyaliao** [comment]
If the issue still exists, please create an issue. We would be glad to resolve it.

**Comment by ORCHORDS** [comment]
The most common reasons Chrome on Android silently drops notifications (even with permission granted) and the canonical fixes:

1. Chrome Android requires the page to be **added to the home screen as a PWA** before notifications work reliably in the background. If your page is just a regular browser tab, notifications fire when the tab is in the foreground but Android's Doze mode and battery optimization suppress background notifications. Verify by:

   - In Chrome on Android, tap the menu (3 dots) → "Add to Home screen".
   - Open the page from the home screen icon (this opens Chrome as a WebAPK, not a regular tab).
   - Trigger the notification from there.

   This is the single most common reason "works on desktop but not on mobile Chrome" — desktop Chrome does not require PWA install, but Android Chrome does for any non-trivial notification behavior.

2. Check the manifest.json fields required for WebAPK installation:

```json
{
  "name": "My App",
  "short_name": "App",
  "start_url": "/",
  "display": "standalone",
  "icons": [
    { "src": "/icon-192.png", "sizes": "192x192", "type": "image/png" },
    { "src": "/icon-512.png", "sizes": "512x512", "type": "image/png" }
  ],
  "gcm_sender_id": "...",
  "gcm_user_visible_only": true
}
```

Missing the `icons` array (with both 192 and 512) is the most common reason Chrome refuses to install the page as a WebAPK, which in turn blocks notifications.

3. Android Chrome requires the notification request to happen **during a user gesture**. This applies to all browsers via the Web Notifications API spec, but Chrome Android is the strictest enforcer:

```ts
import { useWebNotification } from '@vueuse/core'

const { isSupported, requestPermission, send } = useWebNotification({
  title: 'Hello',
  body: 'World',
  icon: '/icon-192.png',         // <-- must be on the same origin
  badge: '/icon-96.png',         // optional, monochrome
  tag: 'unique-tag',             // <-- coalescing id; same tag replaces
  renotify: true,                // <-- vibrate again if replaced
  requireInteraction: false,
  silent: false,
  vibrate: [200, 100, 200],
})

async function onUserClick() {
  if (!isSupported.value) return
  if (Notification.permission === 'default') {
    await requestPermission()
  }
  if (Notification.permission === 'granted') {
    send({ /* can override fields per-call */ })
  }
}
```

If you call `requestPermission()` from inside an `onMounted` hook or some async callback, Android Chrome rejects the request silently. The permission request MUST come from a synchronous user-gesture handler. The click handler in the example above is the right path.

4. Background notifications require ServiceWorker registration. If you want notifications to fire when the tab is not in the foreground, register a service worker and use the service worker's registration to show notifications:

```ts
// Register the service worker once at app start
if ('serviceWorker' in navigator) {
  navigator.serviceWorker.register('/sw.js')
}

// Then use registration.showNotification instead of new Notification
async function showNotification(title: string, body: string) {
  const reg = await navigator.serviceWorker.getRegistration()
  if (reg) {
    await reg.showNotification(title, {
      body,
      icon: '/icon-192.png',
      badge: '/icon-96.png',
    })
  }
}
```

For useWebNotification specifically, the issue is that it calls `new Notification(...)`, which in Chrome on Android can fail silently when the page is backgrounded or not installed as a PWA. The VueUse composable does not fall back to the service worker registration path automatically.

5. Verify by going to chrome://inspect on desktop Chrome with Android Chrome connected via USB. Open the JavaScript console on the Android tab and check what the actual Notification constructor throws. The most common error is:

> "Failed to construct 'Notification': Illegal constructor. Use ServiceWorkerRegistration.showNotification() instead."

Or:

> "Notification permission denied"

Both are silently swallowed by VueUse's useWebNotification. If you want to surface the error, wrap the send call:

```ts
try {
  send({ /* options */ })
} catch (e) {
  console.error('[notify] send failed:', e)
  // Fall back to service-worker-based notification
  if ('serviceWorker' in navigator) {
    const reg = await navigator.serviceWorker.getRegistration()
    reg?.showNotification('Title', { body: 'Body' })
  }
}
```

6. For testing on desktop Chrome that simulates mobile behavior:

   - Open DevTools → toggle device toolbar (Ctrl+Shift+M).
   - The page is now a mobile viewport, but Chrome Desktop still grants notifications normally.
   - To test the actual Android Chrome restrictions, use a real Android device or ChromeOS with the WebAPK developer flags enabled.

For your specific "granted permission, works on desktop, not on mobile" symptom, in order of likelihood:

1. PWA not installed as WebAPK (most common).
2. Manifest missing icons (causes #1).
3. Notification triggered from non-user-gesture (silent rejection).
4. Service worker not registered (background notifications fail).
5. UseWebNotification's `new Notification()` failing silently in Android Chrome (use the service-worker fallback path).

Cross-references:
- vueuse/vueuse#3978 (this thread)
- MDN Web Notifications API: https://developer.mozilla.org/en-US/docs/Web/API/Notifications_API
- Web App Manifest docs: https://web.dev/add-manifest/
- Chrome WebAPK docs: https://developer.chrome.com/docs/android/custom-tabs/integration-guide
- Service worker registration docs: https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API
